app-jee-jsp/src/main/java/org/keycloak/quickstart/appjee/ServiceLocator.java

```diff
@@ -23,12 +23,13 @@ public static URL getServiceUrl(HttpServletRequest req) {
                 return new URL(uri);
             }
 
-            String host = new URL(req.getRequestURL().toString()).getHost();
-            if (host == null) {
-                host = "localhost";
-            }
+            URL requestUrl = new URL(req.getRequestURL().toString());
+
+            String host = requestUrl.getHost();
+            String schema = requestUrl.getProtocol();
+            String port = requestUrl.getPort() != -1 ? String.valueOf(requestUrl.getPort()) : "8080";
```

Comment:
Hmm good catch.

Replacing just context path with /service requires parsing the url - finding the third '/' which may also not be there, so it would then have to be something like:

```
StringBuilder url = new StringBuilder();
String [] parts = getRequestURL().toString().split("/");
for (int i = 0; i < 3 && i < parts.length; i++) {
   url.append(parts[i]).append("/");
}
url.append("service");
```

It seems a less readable code to me. And the only thing I can imagine going wrong with current approach would be a use of user:pass@ in the url, but we don't use that. Am I missing something?